### PR TITLE
[ConstraintSystem] Record generic fix if destination of a pointer con…

### PR DESCRIPTION
**Pull request by xedin**
…version is invalid

If the right-hand side (destination) of value-to-pointer conversion
is incorrect e.g. base type of member is a hole, let's record
a generic "invalid conversion" failure.

Resolves: rdar://problem/68254165

### CODE REVIEW

File: test/Constraints/valid_pointer_conversions.swift [hunk 1]
```diff
@@ -41,3 +41,9 @@ func SR12382(_ x: UnsafeMutablePointer<Double>??) {}
 var i = 0
 SR12382(&i) // expected-error {{cannot convert value of type 'UnsafeMutablePointer<Int>' to expected argument type 'UnsafeMutablePointer<Double>'}}
 // expected-note@-1 {{arguments to generic parameter 'Pointee' ('Int' and 'Double') are expected to be equal}}
+
+//problem/68254165 - Bad diagnostic when using String init(decodingCString:) with an incorrect pointer type
+func rdar68254165(ptr: UnsafeMutablePointer<Int8>) {
+  _ = String(decodingCString: ptr, as: .utf8) // expected-error {{generic parameter 'Encoding' could not be inferred}}
+  // expected-error@-1 {{type '_.Type' has no member 'utf8'}}
```
Comment (xedin) [inline test/Constraints/valid_pointer_conversions.swift:48]: New `HoleType` allows us to deal with this issue and avoid producing member diagnostic here, I'm going to implement that in a follow-up PR to separate concerns here...

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

**Review by hborla — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

**Comment by xedin** [comment]
@swift-ci please smoke test Windows platform

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

**Comment by xedin** [comment]
Windows failure is caused by https://github.com/apple/swift/pull/33306

**Pull request merged**